addDropIndexDesign.md

```diff
@@ -0,0 +1,48 @@
+# Add/Drop Index
+
+## Overview
+
+>*What motivates this to be implemented? What will this component achieve?*
+
+Currently, the system just creates the index without populating it. This component will support actually adding the index (including populating the index) and/or dropping in a transactionally consistent manner. This will be achieved by halting any ongoing transactions and maintaining a delta storage while the index is being built, applying the delta changes after the index is done being built. The bonus task would include index building in a multi-threaded fashion.
+
+## Scope
+>*Which parts of the system will this feature rely on or modify? Write down specifics so people involved can review the design doc*
+
+The parts of the system this feature will modify include the following files:
+ - index_builder.\*
+ - ddl_executors.\*
+ - catalog_accessor.\*
+ - index_schema.\*
+ - codegen.\*
+
+
+## Architectural Design
+>*Explain the input and output of the component, describe interactions and breakdown the smaller components if any. Include diagrams if appropriate.*
+
+There are two phases to index creation. In the first phase, the index is created and added to the catalog (with a flag indicating it is not yet ready for use). Then, that transaction will populate the index as if it were a snapshot of the table at the timestamp of the transaction. _Are we using the standard delta store here or are we maintaining a separate delta store_?
+
+At the start of the next phase (which starts at t2), a separate flag is set in the catalog - indicating that any transactions with a timestamp after t2 that wish to modify the column the index is being built on must block. The transaction that is building the index then blocks - waiting for all transactions with timestamp less than t1 to commit. Once that is done, the index building transaction scans through the delta store and updates the index with any changes.
```

Comment:
what about transactions with timestamp > t2 that access the column but do not modify the column? Should they be blocked?

Reply:
Are there other possibilities apart from blocking? Is it possible to fall back to a sequential scan?

Reply:
Also, I though indexes can be built on any collection of tupleslots, not necessarily just columns? For example, I can have an index on a table.